docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
 
 * It helps facilitate non-disruptive P4 program updates.
 
-The shortest possible binary string for an integer `value` with `high_bit`
-as the most-significant non-zero bit in the value is computed as:
+In particular, the kinds of P4 program updates that this
+representation facilitates are those where a P4Runtime server and
+client can continue to transmit P4Runtime messages between them when
+one has a P4Info file for version A of a P4 program, at the same time
+that the other has a P4Info file for version B of a P4 program, and
+those P4 programs differ in the bit widths of some values of type
+`bit<W>` and/or `int<W>`.
+
+Note that this representation does _not_ make it possible to
+seamlessly change the type of a value from signed to unsigned, or vice
+versa.  If you attempt to do so, this mechanism can quietly change
+negative signed values to positive unsigned values, or vice versa.  It
+also limits the magnitude of the values transmitted to those that fit
+within the smaller of the bit widths supported by either end of the
+message transmission.  If a message sender attempts to send a value
+larger than the receiver expects, the receiver will detect it as out
+of range.
+
+In the P4Runtime API version 1.0, values of table key fields, action
+parameters, and fields in packet-in and packet-out headers between a
+device and the controller (see [#sec-controller-packet-meta]), may not
+be of type `int<W>`.  The rules for encoding signed values thus only
+apply to messages of type `P4Data` (see
+[#sec-p4data-in-p4runtime-proto]).
+
+For a value of type `bit<W>`, the fewest number of bits required to
+represent the integer value $V > 0$ is the smallest integer $A$ such
+that $V \leq 2^A - 1$.
+
+For a value of type `int<W>`, the fewest number of bits required to
+represent the integer value $V \neq 0$ in 2's complement form is the
+smallest integer $A$ such that $-2^{A-1} \leq V \leq 2^{A-1} - 1$.
+
+As a special case, define that the value $V=0$ requires at least $A=1$
+bit to represent, regardless of whether it is signed or unsigned.
+
+The shortest possible binary string for an integer $V$ that needs $A$
+bits to represent it is computed as:
 ~ Begin CPP
-if (value != 0)
-  encoded_string_size = (high_bit + 8) / 8;
-else
-  encoded_string_size = 1;
+encoded_string_size = floor((A + 7) / 8)
```

Comment:
From teverman: Would it be clearer to rename encoded_string_size to minimum_string_size? Just a thought.

Reply:
Done.